core/src/main/scala/org/apache/spark/executor/ExecutorMetricsPoller.scala

```diff
@@ -124,17 +124,11 @@ private[spark] class ExecutorMetricsPoller(
    */
   def onTaskCompletion(taskId: Long, stageId: Int, stageAttemptId: Int): Unit = {
     // Decrement the task count.
-    // Remove the entry from stageTCMP if the task count reaches zero.
 
     def decrementCount(stage: StageKey, countAndPeaks: TCMP): TCMP = {
       val countValue = countAndPeaks.count.decrementAndGet()
```

Comment:
Shall we add `assert(countValue >= 0)` here to ensure the task count doesn't go wrong? And then in `removeIfInactive`, we could use `v.count.get == 0`.

Reply:
Yeah that sounds good, I added it.

Reply:
@Ngone51 we should revisit the assert as 
https://issues.apache.org/jira/browse/SPARK-34779?focusedCommentId=17610965&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17610965

Reply:
Ok, I have a theory. Let me run some experiment.

Reply:
So the basic assumption to have at least as many `onTaskStart()` calls s `onTaskCompletion()` for the same `stageId` & `stageAttemptId` pair is correct!
But still there can be a race condition between `getExecutorUpdates()` and `onTaskStart`.

Let me show you one of the possible scenario. 

So assuming we are after a finished a task which was running on its own on this executor 
so it decreases the `count` from 1 to 0.

On the task runner thread let say a new task starts. 
So we are in the `onTaskStart()` method we already taken out the `TCMP` 
where the counter is 0 but we are just before incrementing the counter:

```
 val countAndPeaks = stageTCMP.computeIfAbsent((stageId, stageAttemptId),
      _ => TCMP(new AtomicLong(0), new AtomicLongArray(ExecutorMetricType.numMetrics)))
 val stageCount = countAndPeaks.count.incrementAndGet()
```

On the heartbeater thread `getExecutorUpdates()` is running to be exact the `removeIfInactive()`:

```
   def removeIfInactive(k: StageKey, v: TCMP): TCMP = {
      if (v.count.get == 0) {
        logDebug(s"removing (${k._1}, ${k._2}) from stageTCMP")
        null
      } else {
        v
      }
    }
```

And we remove the entry from `stageTCMP` as the count is 0.

Lets go back to the task runner thread where we increase the counter to 1 but that value will be lost 
as we have no entry in the `stageTCMP` for this stage and attempt.

So if a new task comes instead of 2 we will have 1 in the `stageTCMP` and when those two tasks finishes 
the second one will decrease the counter from 0 to -1.

Reply:
I have a fix in my mind but let's wait for the logs to verify my theory...